From cf5163d314c3ef403f79b2d83a673959e259545b Mon Sep 17 00:00:00 2001
From: Thomas Orgis <thomas@orgis.org>
Date: Mon, 22 Jan 2018 16:46:38 +0100
Subject: [PATCH] Create git diff files with indicated mode.

This enables correct mode for newly created files as indicated in
git diffs. The lack of this will hurt people trying to build current
Linux 4.14.x from the signed tarball and patch file, as it creates
a script which breaks the build if it is not created with the executable
bit.
---
 src/patch.c             |  8 ++++++--
 tests/Makefile.am       |  1 +
 tests/file-create-modes | 53 +++++++++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 60 insertions(+), 2 deletions(-)
 create mode 100644 tests/file-create-modes

diff --git a/src/patch.c b/src/patch.c
index bba7e0e..0120aa6 100644
--- a/src/patch.c
+++ b/src/patch.c
@@ -590,8 +590,12 @@ main (int argc, char **argv)
 			}
 
 		      if (inerrno)
-			set_file_attributes (TMPOUTNAME, attr, NULL, NULL,
-					     mode, &new_time);
+			{
+			  if(set_mode)
+			    attr |= FA_MODE; /* Create files with proper mode given in git diffs. */
+			  set_file_attributes (TMPOUTNAME, attr, NULL, NULL,
+					       mode, &new_time);
+			}
 		      else
 			{
 			  attr |= FA_IDS | FA_MODE | FA_XATTRS;
diff --git a/tests/Makefile.am b/tests/Makefile.am
index 9ae6916..6b6df63 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -35,6 +35,7 @@ TESTS = \
 	empty-files \
 	false-match \
 	fifo \
+	file-create-modes \
 	file-modes \
 	filename-choice \
 	git-binary-diff \
diff --git a/tests/file-create-modes b/tests/file-create-modes
new file mode 100644
index 0000000..7f3c9e8
--- /dev/null
+++ b/tests/file-create-modes
@@ -0,0 +1,53 @@
+# Copyright (C) 2010-2018 Free Software Foundation, Inc.
+#
+# Copying and distribution of this file, with or without modification,
+# in any medium, are permitted without royalty provided the copyright
+# notice and this notice are preserved.
+
+# Test if git diffs can create filess with proper modes
+
+. $srcdir/test-lib.sh
+
+require cat
+require sed
+use_local_patch
+use_tmpdir
+
+# ==============================================================
+
+cat > f.diff <<EOF
+diff --git a/f b/f
+new file mode 100710
+index 0000000..3e75765
+--- /dev/null
++++ b/f
+@@ -0,0 +1 @@
++new
+
+diff --git a/g b/g
+new file mode 100654
+index 0000000..3e75765
+--- /dev/null
++++ b/g
+@@ -0,0 +1 @@
++new
+EOF
+
+check 'patch -p1 < f.diff' <<EOF
+patching file f
+patching file g
+EOF
+
+check 'ls -l f g | sed "s,\(..........\).*,\1,"' <<EOF
+-rwx--x---
+-rw-r-xr--
+EOF
+
+check 'patch -p1 -R < f.diff' <<EOF
+patching file f
+patching file g
+EOF
+
+check 'test -e f || test -e g && echo fail || echo pass' <<EOF
+pass
+EOF
-- 
2.15.1

